pkg/hubble/filters/workload.go

```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Hubble
+
+package filters
+
+import (
+	"context"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+	v1 "github.com/cilium/cilium/pkg/hubble/api/v1"
+)
+
+func filterByWorkload(workloads []*flowpb.Workload, getEndpoint func(*v1.Event) *flowpb.Endpoint) FilterFunc {
+	return func(ev *v1.Event) bool {
+		for _, w := range getEndpoint(ev).GetWorkloads() {
+			for _, f := range workloads {
+				if (f.GetName() == "" || f.GetName() == w.GetName()) &&
```

Comment:
This returns true when both Name and Kind are both empty, which seems like that should match nothing, ie: if I want to filter for flows without a workload. I also notice that this case isn't in the tests.

Reply:
> This returns true when both Name and Kind are both empty, which seems like that should match nothing

Could you please elaborate why you think it should match nothing? What's the value of a filter than doesn't match anything? I'm not saying that a filter than match everything (i.e. the current patch behaviour) is useful, but at least it is consistent with the idea of "empty string means wildcard" for the kind/name workload fields.

> I also notice that this case isn't in the tests.

Thanks it's missing a test case indeed!

Reply:
Oh that's a great point, as it stand there is no way to express "match pods without workloads" which I guess could be useful.

Now implementation wise, providing an repeated value when you want to match no value feels strange. I think we should wrap the struct instead.

Reply:
Added a `WorkloadFilter` message wrapping the `Workload` list. Providing a `WorkloadFilter` with an empty `Workload` list will match flows without workloads, see `{source,destination}-filter-empty{,-match}` test cases.

Reply:
I have to say I do find the new design a bit inconsistent with all other existing flow filters. The wrapper is a new concept that does not exist for any other filter, which I find a bit odd.

If I understand correctly, the goal is to be able to the following (I'll ignore case 0 in the discussion, because it's implementation seems straightforward):

0. Query flows with a particular workload name, namespace, or both ("normal" filters)
1. Query flows regardless of workload (the default, if no filter specified)
2. Query flows only flows with a non-empty workload
3. Query flows only flow with an empty/missing workload

The current API seems to implement case 1 with an absent wrapper, case 2 with a wrapper containing a workload list with only empty fields, case 3 with a wrapper with an empty workload list.

However, it seems that case 3 matches exactly the opposite set of flows as in case 2. To me, means that it should be done via an exclude-filter, i.e. have case 3 implemented as an exclude-filter matching all flows with an empty workload.

My proposal therefore would be to have get rid of the wrapper (which again, we don't have for any other filters either), and implement the cases as follows:

Bring back the plain repeated list:
```
repeated Workload {source,destination}_workload
```

Case 1: Implement via empty `{source,destination}_workload` (nothing special here, same as we do for all other filters)
Case 2: Implement via a *include*-list `{source,destination}_workload` the list containing the empty workload `{source,destination}_workload = {Name: "", Namespace ""}`
Case 3: Implement *exclude*-list `{source,destination}_workload` the list containing the empty workload `{source,destination}_workload = {Name: "", Namespace ""}`

For consistency with other filters, it seems that the pod filter disallows having both an empty name and namespace. The node filter however implements the same behavior as proposed here as well, where both an empty clustername and nodename only match non-empty node-names:

https://github.com/cilium/cilium/blob/1108684e354d2b6ecf66cbaf1cd82de5fa88b70f/pkg/hubble/filters/nodename_test.go#L42

Reply:
@gandro thanks for the explanation, all you wrote make sense I wasn't able to wrap my mind (no pun intended) around exclude before refactoring the patch hence the wrapper change. What I'll do is revert to the `repeated Workload` list in the filter and keep all the tests cases using the proposed approach.

Reply:
I reworked the patch and testing is now split into two `func`, one using the filter as include list and the other `TestWorkloadFilterExclude` using the filter as exclude list).